src/apps/finance/hooks/useBalances.js

```diff
@@ -0,0 +1,130 @@
+import { useEffect, useMemo, useState } from 'react'
+import { useConnect } from '@1hive/connect-react'
+import { useOrganizationState } from '@/providers/OrganizationProvider'
+import { useContractReadOnly, getContract } from '@/hooks/shared/useContract'
+import { getAppByName } from '@/utils/app-utils'
+import BN from 'bn.js'
+import usePromise from '@/hooks/shared/usePromise'
+// import { toMs } from '@/utils/date-utils'
+
+import vaultBalanceAbi from '../abi/vault-balance.json'
+import minimeTokenAbi from '@/abi/minimeToken.json'
+import { useMounted } from '@/hooks/shared/useMounted'
+
+const VAULT_NAME = 'agent'
+const INITIAL_TIMER = 2000
+
+const CHAIN_TOKEN_ADDRESS = '0x0000000000000000000000000000000000000000'
+
+const useBalances = (timeout = 7000) => {
+  const { apps, connectedFinanceApp } = useOrganizationState()
+  const [tokenBalances = [], { loading, error }] = useConnect(
+    () => connectedFinanceApp?.onBalance(),
+    [connectedFinanceApp]
+  )
+  const mounted = useMounted()
+  const [tokenData, setTokenData] = useState([])
+  const [tokenWithBalance, setTokenWithBalance] = useState([])
+  const [loadingBalances, setLoadingBalances] = useState(true)
+
+  const chainId = 100 // TODO HANDLE CHAINS
+  const vaultContract = useContractReadOnly(
+    getAppByName(apps, VAULT_NAME).address,
+    vaultBalanceAbi,
+    chainId
+  )
+
+  useEffect(() => {
+    if (!tokenBalances?.length || !mounted) {
+      return
+    }
+    const getTokenData = async () => {
+      try {
+        const tokensWithData = await Promise.all(
+          tokenBalances.map(async tokenBalance => {
+            const tokenContract = getContract(
```

Comment:
We're instantiating a new token ethers `Contract` every 5s as a result of the connector token balances polling. They're a bit expensive to create so I'm thinking maybe it would better to cache them  in a external object/map and  re-use them in every run. Wdyt?